addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/discovery/SqueezeBoxServerDiscoveryService.java

```diff
@@ -0,0 +1,275 @@
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.squeezebox.discovery;
+
+import static org.openhab.binding.squeezebox.SqueezeBoxBindingConstants.*;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.squeezebox.SqueezeBoxBindingConstants;
+import org.openhab.binding.squeezebox.internal.utils.HttpUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SqueezeBoxServerDiscoveryService sends a UDP message on port 3483 to the
+ * network in order to trigger listening Squeeze Servers to advertise
+ * themselves.
+ * 
+ * @author Dan Cunningham
+ *
+ */
+public class SqueezeBoxServerDiscoveryService extends AbstractDiscoveryService {
+	private final Logger logger = LoggerFactory
+			.getLogger(SqueezeBoxServerDiscoveryService.class);
+
+	/**
+	 * Give us 5 seconds to find servers
+	 */
+	private static int TIMEOUT = 5;
+	/**
+	 * Port SqueezeServers listen on for discovery requestd
+	 */
+	private static final int DISCO_PORT = 3483;
+	/**
+	 * Discovery packet we send on the network to find Squeeze Servers
+	 */
+	private String discoPacket = "eIPAD\000NAME\000JSON\000VERS\000UUID\000JVID\006\012\034\056\078\012\034";
+	/**
+	 * Discovery replay packet containing the name, web port, version and UUID
+	 * of a Squeeze Server
+	 */
+	private Pattern replyPattern = Pattern
+			.compile("ENAME(.*?)JSON(.*?)VERS(.*?)UUID\\$(.*?)");
+	/**
+	 * When true we will keep sending out packets
+	 */
+	private boolean discoveryRunning;
+	/**
+	 * our Discovery socket
+	 */
+	private DatagramSocket socket;
+
+	/**
+	 * Looks for Squeeze Servers on the local network
+	 */
+	public SqueezeBoxServerDiscoveryService() {
+		super(Collections.singleton(SQUEEZEBOXSERVER_THING_TYPE), TIMEOUT, true);
+	}
+
+	@Override
+	public Set<ThingTypeUID> getSupportedThingTypes() {
+		return Collections.singleton(SQUEEZEBOXSERVER_THING_TYPE);
+	}
+
+	@Override
+	protected void startScan() {
+		scanServer();
+	}
+
+	@Override
+	protected void startBackgroundDiscovery() {
```

Comment:
this method should actually return fast. If I understand the code right, then "scanServer" will actually block while discovery is running.
Furthermore, it probably does not make much sense to wait forever for responses when no new broadcast is sent out. So maybe you should schedule a regular job that runs "scanServer()" every couple of minutes?